src/main/java/com/hierynomus/smbj/session/Session.java

```diff
@@ -143,6 +143,25 @@ private Share connectTree(final String shareName) {
             Future<SMB2TreeConnectResponse> send = this.send(smb2TreeConnectRequest);
             SMB2TreeConnectResponse response = Futures.get(send, connection.getConfig().getTransactTimeout(), TimeUnit.MILLISECONDS, TransportException.Wrapper);
             if (response.getHeader().getStatus().isError()) {
+                if (dfsPathResolver != null) {
```

Comment:
The spec is kind of vague on this. The flowchart in 3.1.4.1 seems to indicate a DFS client should actually do referral queries proactively.
In this case I got BAD_NETWORK_NAME from a Windows server (i.e., share does not exist). Since it's not specified in the spec I decided to just try on any error thinking it can't do much harm and this you're not dependent on what might be implementation specific behaviour.

Reply:
Yes, the pro-active approach might be indeed what the flowchart is hinting at.
It's a bit spread through the code now. Let me think how to improve this.

Reply:
Not sure if that's necessary; I was just trying to explain that the spec doesn't really say which error codes should trigger a DFS referral query.
FWIW on my mac I see plain open requests going over the wire. The DFS referral requests only happen after a `PATH_NOT_COVERED` error response.
Querying the cache prior to the open request seems like a good enhancement though. I.e., query cache, if cache hit use cache, otherwise do a regular open and then do the DFS referral request on `PATH_NOT_COVERED` errors.